File: src/go/problems/summary_ranges_test.go
// https://leetcode.com/problems/summary-ranges

package problems

import (
	"fmt"
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type SummaryRanges struct{}

func TestSummaryRanges(t *testing.T) {
	gen := core.TestSuite[SummaryRanges]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([]int{0, 1, 2, 4, 5, 7}).Result([]string{"0->2", "4->5", "7"})
	}).Add(func(tc *core.TestCase) {
		tc.Param([]int{0, 2, 3, 4, 6, 8, 9}).Result([]string{"0", "2->4", "6", "8->9"})
	}).Run(t)
}

func (SummaryRanges) Solution(nums []int) []string {
	result := []string{}
	startAt := 0
	length := len(nums)
	for i := 0; i < length; i++ {
		if i == length-1 || nums[i]+1 < nums[i+1] {
			if i == startAt {
				result = append(result, fmt.Sprintf("%d", nums[startAt]))
			} else {
				result = append(result, fmt.Sprintf("%d->%d", nums[startAt], nums[i]))
			}

			startAt = i + 1
		}
	}

	return result
}

File: src/python/problems/summary_ranges_test.py
# https://leetcode.com/problems/summary-ranges

from core.problem_base import *

class SummaryRanges(ProblemBase):
    def Solution(self, nums: list[int]) -> list[str]:
        result = []
        startAt = 0
        length = len(nums)
        for i in range(length):
            if i == length - 1 or nums[i] + 1 < nums[i + 1]:
                if i == startAt:
                    result.append(f"{nums[startAt]}")
                else:
                    result.append(f"{nums[startAt]}->{nums[i]}")

                startAt = i + 1

        return result

if __name__ == '__main__':
    TestGen(SummaryRanges) \
        .Add(lambda tc: tc.Param([0,2,3,4,6,8,9]).Result(["0","2->4","6","8->9"])) \
        .Add(lambda tc: tc.Param([0,1,2,4,5,7]).Result(["0->2","4->5","7"])) \
        .Run()
